textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/GramBooleanQuery.java

```diff
@@ -293,6 +294,133 @@ private static String toLuceneQueryString(GramBooleanQuery query) {
 			}
 		}
 	}
+	
+	public boolean isEmpty() {
+		if (this.operandSet.size() > 0) {
+			return false;
+		}
+		for (GramBooleanQuery subQuery : this.subQuerySet) {
+			if (! subQuery.isEmpty()) {
+				return false;
+			}
+		}
+		return true;
+	}
+	
+	// "AND" two DNF tree
+	// Apply distributive laws: 
+	// (a OR b) AND (c OR d) --> (a AND c) OR (a AND d) OR (b AND c) OR (c AND d)
+	private static GramBooleanQuery andDNF(GramBooleanQuery left, GramBooleanQuery right) {
+		if (left.isEmpty()) {
+			return right;
+		}
+		if (right.isEmpty()) {
+			return left;
+		}
+		GramBooleanQuery resultQuery = new GramBooleanQuery(QueryOp.OR);
+		for (String leftOperand : left.operandSet) {
+			for (String rightOperand : right.operandSet) {
+				GramBooleanQuery tempQuery = new GramBooleanQuery(QueryOp.AND);
+				tempQuery.operandSet.add(leftOperand);
+				tempQuery.operandSet.add(rightOperand);
+				resultQuery.subQuerySet.add(tempQuery);
+			}
+			for (GramBooleanQuery rightSubQuery : right.subQuerySet) {
+				GramBooleanQuery tempQuery = new GramBooleanQuery(QueryOp.AND);
+				tempQuery.operandSet.add(leftOperand);
+				tempQuery.operandSet.addAll(rightSubQuery.operandSet);
+				resultQuery.subQuerySet.add(tempQuery);
+			}
+		}
+		for (GramBooleanQuery leftSubQuery : left.subQuerySet) {
+			for (String rightOperand : right.operandSet) {
+				GramBooleanQuery tempQuery = new GramBooleanQuery(QueryOp.AND);
+				tempQuery.operandSet.addAll(leftSubQuery.operandSet);
+				tempQuery.operandSet.add(rightOperand);
+				resultQuery.subQuerySet.add(tempQuery);
+			}
+			for (GramBooleanQuery rightSubQuery : right.subQuerySet) {
+				GramBooleanQuery tempQuery = new GramBooleanQuery(QueryOp.AND);
+				tempQuery.operandSet.addAll(leftSubQuery.operandSet);
+				tempQuery.operandSet.addAll(rightSubQuery.operandSet);
+				resultQuery.subQuerySet.add(tempQuery);
+			}
+		}
+		return resultQuery;
+	}
+	
+	// After Transforming to DNF, apply Absorption laws to simplify it
+	// a OR (a AND b) --> a
+	// Tree must be already transformed to DNF before calling this function!
+	public static GramBooleanQuery simplifyDNF(GramBooleanQuery query) {
```

Comment:
This function checks all sub-queries and see if this queries is replaceable by another one. The final result will only contain necessary queries.